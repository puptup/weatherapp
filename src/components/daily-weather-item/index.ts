export { DailyWeatherItem } from "./daily-weather-item";
