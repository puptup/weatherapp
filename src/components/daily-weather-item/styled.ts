import { device } from "@constants/devices";
import styled from "styled-components";

type WeatherItemProps = {
  active: boolean;
};

export const WeatherItem = styled.div<WeatherItemProps>`
  position: relative;
  width: 230px;
  padding: ${({ theme }) => theme.spacing.padding};
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  align-items: center;
  border-radius: ${({ theme }) => theme.spacing.borderRadiusItem};
  background-color: ${({ active, theme }) =>
    active ? theme.palette.item.active : theme.palette.item.primary};
  margin-bottom: ${({ theme }) => theme.spacing.betweenItems};
  transition: all 0.15s linear;
  ${({ active, theme }) =>
    !active &&
    `
    cursor: pointer;
    &:hover {
      background-color: ${theme.palette.item.hover};
    }
  
    &:active {
      background-color: ${theme.palette.item.active};
    }
  `}

  @media ${device.laptop} {
    width: 60px;
    display: flex;
    flex-direction: column;
    margin: 5px;
    flex-shrink: 0;
  }
`;

export const DateWrapper = styled.div``;

export const AverageTemp = styled.span``;
export const MinMaxTemp = styled.span``;

export const Image = styled.img`
  display: block;
  margin: 0 auto;
  width: 40px;
  height: 40px;
`;

export const TempWrapper = styled.div`
  transition: all 0.15s linear;
  text-align: center;
`;
