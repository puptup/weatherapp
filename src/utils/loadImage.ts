const comparer = {
  "01d": 2,
  "01n": 26,
  "02d": 3,
  "02n": 27,
  "03d": 7,
  "03n": 7,
  "04d": 5,
  "04n": 29,
  "09d": 13,
  "09n": 32,
  "10d": 11,
  "10n": 11,
  "11d": 14,
  "11n": 14,
  "13d": 19,
  "13n": 34,
  "50d": 9,
  "50n": 9,
};

type ComapereKeys = keyof typeof comparer;

const getOpenWeatherIcon = (iconID: string) => {
  const newIconID = comparer[iconID as ComapereKeys];
  return newIconID || 1;
};

export const loadIcon = async (iconID: string, openWeather: boolean = false) => {
  if (openWeather) {
    return import(`../assets/weather-icons/${getOpenWeatherIcon(iconID)}.png`);
  }
  return import(`../assets/weather-icons/${iconID}.png`);
};
