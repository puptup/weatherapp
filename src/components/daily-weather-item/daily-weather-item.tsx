import { useAppDispatch } from "@hooks";
import { setDayAction } from "@store/actions-creators";
import { setTheme } from "@store/reducers/theme-slice";
import { DailyWeather } from "@types";
import { getWeekdayOrToday } from "@utils/date";
import { loadIcon } from "@utils/loadImage";
import { getThemeByIcon } from "@utils/theme-by-icon";
import React, { useEffect, useState } from "react";

import { AverageTemp, DateWrapper, Image, TempWrapper, WeatherItem } from "./styled";

type DailyWeatherItemProps = {
  item: DailyWeather;
  currentDay: boolean;
};

export const DailyWeatherItem = ({ item, currentDay }: DailyWeatherItemProps) => {
  const [image, setImage] = useState<string>();
  const dispatch = useAppDispatch();

  const { icon, date, temperature } = item;

  useEffect(() => {
    loadIcon(icon).then((path) => {
      setImage(path.default);
    });
  }, []);

  const handleDay = () => {
    if (!currentDay) {
      dispatch(setDayAction(new Date(date)));
      const newTheme = getThemeByIcon(icon);
      dispatch(setTheme(newTheme));
    }
  };

  return (
    <WeatherItem active={currentDay} onClick={handleDay} data-testid="daily-weather-item">
      <DateWrapper>{getWeekdayOrToday(new Date(date), true)}</DateWrapper>
      <Image src={image} alt="weather icon" />
      <TempWrapper>
        <AverageTemp>{Math.round(temperature)}</AverageTemp>
      </TempWrapper>
    </WeatherItem>
  );
};
