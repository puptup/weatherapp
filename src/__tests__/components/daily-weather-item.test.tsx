import "@testing-library/jest-dom";

import { DailyWeatherItem } from "@components/daily-weather-item/daily-weather-item";
import * as utils from "@utils/loadImage";
import { renderWithProviders } from "@utils/testing/render-with-providers";
import React from "react";

describe("hourly-weather-item", () => {
  jest.spyOn(utils, "loadIcon").mockReturnValue(Promise.resolve(() => "test"));

  it("should correspond to snapshot", () => {
    const { getByTestId } = renderWithProviders(
      <DailyWeatherItem
        item={{
          icon: "test",
          date: "2023-01-23T14:13:22.224Z",
          maxTemperature: 0,
          minTemperature: 0,
          temperature: 0,
          weather: "test",
        }}
        currentDay
      />
    );
    expect(getByTestId("daily-weather-item")).toMatchSnapshot();
  });
});
