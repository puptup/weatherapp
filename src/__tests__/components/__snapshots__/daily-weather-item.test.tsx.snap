// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hourly-weather-item should correspond to snapshot 1`] = `
<div
  class="sc-bcXHqe iWCNhl"
  data-testid="daily-weather-item"
>
  <div
    class="sc-gswNZR"
  >
    Today
  </div>
  <img
    alt="weather icon"
    class="sc-eDvSVe gyrvxY"
  />
  <div
    class="sc-jSUZER eQHxsh"
  >
    <span
      class="sc-dkrFOg"
    >
      0
    </span>
  </div>
</div>
`;
